### PR TITLE
Kore.Reachability: decouple result type of applyClaims and applyAxioms

### PR DESCRIPTION
**Pull request by ana-pantilie**
Fixes https://github.com/kframework/kore/issues/2195

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: kore/src/Kore/Reachability/Claim.hs [hunk 1]
```diff
@@ -336,24 +332,29 @@ transitionRule claims axiomGroups = transitionRuleWorker
...
+    transitionRuleWorker ApplyClaims (Claimed claim) = do
+        applied <- applyClaims claims claim
+        case applied of
+            -- If the result is \\bottom, the claim is proven on the
+            -- current branch.
+            ReachedBottom -> return Proven
+            ApplyRewritten appliedClaim ->
+                return (Rewritten appliedClaim)
+            ApplyRemainder appliedClaim ->
+                return (Remaining appliedClaim)
```
Comment (ana-pantilie) [inline kore/src/Kore/Reachability/Claim.hs:344]: Do we want to factor this out? Or do we want to keep the state transitions as explicit as possible in the body of `transitionRule`?

File: kore/src/Kore/Reachability/Claim.hs [hunk 2]
```diff
@@ -336,24 +332,29 @@ transitionRule claims axiomGroups = transitionRuleWorker
...
-    -- results in `derivePar` as opposed to here.
-
-    transitionRuleWorker ApplyClaims (Claimed claim) =
-        applyClaims claims claim
+    transitionRuleWorker ApplyClaims (Claimed claim) = do
+        applied <- applyClaims claims claim
+        case applied of
+            -- If the result is \\bottom, the claim is proven on the
+            -- current branch.
+            ReachedBottom -> return Proven
```
Comment (ttuegel) [inline kore/src/Kore/Reachability/Claim.hs:340]: This is already handled in `checkImplication`. Why do we need to handle it here, too?

File: kore/src/Kore/Reachability/Claim.hs [hunk 3]
```diff
@@ -167,6 +168,13 @@ import Pretty
     )
 import qualified Pretty
 
+data ApplyResult claim
+    = ApplyRewritten !claim
+    | ApplyRemainder !claim
+    | ReachedBottom
```
Comment (ttuegel) [inline kore/src/Kore/Reachability/Claim.hs:174]: Please give this some documentation.

File: kore/src/Kore/Reachability/Claim.hs (hunk 1 above)
Comment (ttuegel) [reply kore/src/Kore/Reachability/Claim.hs:344]: I think we should factor this out.

**Review by ttuegel — APPROVED** [review]

**Pull request merged**